renderlib/graphics/gl/Font.cpp

```diff
@@ -0,0 +1,146 @@
+#include "Font.h"
+
+#include "Util.h"
+
+#define STB_TRUETYPE_IMPLEMENTATION
+#include "stb/stb_truetype.h"
+
+#include <stdio.h>
+
+static constexpr int s_textureSize = 512;
+
+GLuint
+generateFontTexture(unsigned char* temp_bitmap)
+{
+  // expand temp_bitmap to RGBA
+  unsigned char* expanded_bitmap = new unsigned char[s_textureSize * s_textureSize * 4];
+  for (int i = 0; i < s_textureSize * s_textureSize; ++i) {
+    expanded_bitmap[i * 4 + 0] = 255;
+    expanded_bitmap[i * 4 + 1] = 255;
+    expanded_bitmap[i * 4 + 2] = 255;
+    expanded_bitmap[i * 4 + 3] = temp_bitmap[i];
+  }
+
+  GLuint ftex = 0;
+  glGenTextures(1, &ftex);
+  glBindTexture(GL_TEXTURE_2D, ftex);
+  glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA8, s_textureSize, s_textureSize, 0, GL_RGBA, GL_UNSIGNED_BYTE, expanded_bitmap);
+  check_gl("load font texture");
+  // can free temp_bitmap at this point
+  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+  check_gl("set filtering on font texture");
+  delete[] expanded_bitmap;
+  return ftex;
+}
+
+Font::Font()
+{
+  m_texID = 0;
+}
+
+Font::~Font()
+{
+  unload();
+}
+
+void
+Font::load(const char* filename)
+{
+  if (m_texID != 0) {
+    return;
+  }
+
+  FILE* file = fopen(filename, "rb");
+  if (!file) {
+    return;
+  }
+
+  fseek(file, 0, SEEK_END);
+  long size = ftell(file);
+  fseek(file, 0, SEEK_SET);
+
+  unsigned char* buffer = (unsigned char*)malloc(size);
+  fread(buffer, 1, size, file);
+  fclose(file);
+
+  unsigned char temp_bitmap[s_textureSize * s_textureSize];
+  stbtt_BakeFontBitmap(buffer,
+                       0,
+                       32.0,
+                       temp_bitmap,
+                       s_textureSize,
+                       s_textureSize,
+                       m_firstChar,
+                       m_numChars,
+                       m_cdata); // no guarantee this fits!
```

Comment:
Scary... If we ever do run over this texture size limit, how will we be able to tell? Will it just cut off out-of-bounds characters or will it start to rewrite parts of the texture array?

Reply:
I think this is one of those things that is boilerplate example code from the font lib I am using, and it was more effort to properly calculate this than the risk of it actually going bad in the general use case.  I'll write a new ticket just to keep it on the table.